### DISCUSSION TITLE
[React] Support other floating element use cases

### DISCUSSION THREAD

**Discussion by atomiks** [floating-ui/floating-ui#2718] [React]
https://github.com/floating-ui/floating-ui/discussions/2191

Since this library is called "Floating UI", we can and should do more than just support elements that use anchor positioning — for example, a non-anchored floating Dialog or Drawer component are already possible to build using the existing interaction primitives, but a component like a Toast or anything draggable isn't despite them also being "floating" elements.

Non-anchored UI elements that also float:

- Dialog (already supported)
- Drawer (already supported)
- Toast
- Draggable window
- etc

**Comment by takurinton** [comment]
@atomiks 
Hi atomiks, Thanks for this great package.
I wanted a toast component using floating-ui and was actually working on a prototype at hand.

source code: https://github.com/takurinton/floating-ui-toast
demo: https://floating-ui-toast.vercel.app
codesandbox: https://codesandbox.io/s/infallible-swanson-fbdx15?file=/src/App.tsx

This prototype is in the prototype stage at this time and has issues with accessibility, state management, and interface.
I am inclined to clean up this prototype a bit more and am actually trying to work on it. 

I am inclined to contribute to floating-ui, can I add this as a demo for the toast component of floating-ui??

**Comment by atomiks** [comment]
Nice start @takurinton, any component should follow the [ARIA design pattern](https://www.w3.org/TR/wai-aria/#aria-live) and these features:

- Pauses closing on hover, focus and window blur
- Supports hotkey to jump to toast viewport 
- Supports closing via swipe gesture

I think an absolute flexbox column is better for stacking than each being individually positioned with a dynamic `top`?

**Comment by takurinton** [comment]
Thanks for taking a look!!

> any component should follow the [ARIA design pattern](https://www.w3.org/TR/wai-aria/#aria-live) and these features:

This I did not take into account in the initial stages. I think these can be solved with hooks that already exist in floating-ui, am I correct?

> I think an absolute flexbox column is better for stacking than each being individually positioned with a dynamic top?

It was a trial-and-error stage, so it is implemented as it is now, but I feel you are right.
This is a little bit of a commit to load a few modifications.

**Comment by atomiks** [comment]
The `useRole` hook doesn't support it because it only references `aria-haspopup` roles, whereas the toast element needs to have this structure:

```js
<ul role="region" aria-live="polite">
  <li role="status" aria-atomic="true">
     Toast content
  </li>
</ul>
```

`useDismiss` works for `esc` dismissal, and `FloatingFocusManager` allows it to be interactive (e.g. to press an Undo button inside of it), though. The pattern isn't very complex however, and should be straightforward to support as seen above.

**Comment by takurinton** [comment]
Thanks support. I am working on implementing the above.  

One question,(in https://github.com/floating-ui/floating-ui/issues/2192#issuecomment-1495005516)

> - Supports closing via swipe gesture

Is this supported by the existing floating-ui hooks? Or do I need to implement my own hooks (e.g. `useDrag`, `useSwipe`, or other name)?

**Reply by kotAPI** [reply]
any luck on this @takurinton ?
